feat(grid): update aspect ratio with graceful overflow

photodow opened the pull request:
@joshblack I figured I could quickly put in a pull request for the updates we were having a sideline about.

Basically this pull request simplifies the aspect ratio implementation on the grid and allows content to gracefully overflow instead of breaking the layout when it happens. This is something I've been exploring in ways of leveraging on IBM.com and was thinking it made more sense inside of Carbon Core. Maybe y'all already looked into it, but I wanted to get y'alls thoughts and see if there are any drawbacks to this method that I might not be considering.

Thanks!

##### As-is
![image](https://user-images.githubusercontent.com/3793636/73695405-a62e0680-469f-11ea-967e-aa11784d368c.png)

##### Example with graceful overflow
![image](https://user-images.githubusercontent.com/3793636/73695290-6bc46980-469f-11ea-81c5-66b1cbfc3a74.png)


#### Changelog

**New**

- Manage aspect ratio
- Added in additional reversed ratios (e.g. instead of `16:9`, `9:16`)

**Changed**

- Divided aspect ratio mixin into a couple of functions and mixins
- Updated `/examples`

**Removed**

- previous implementation, and extra code left over from it.

photodow committed:
chore(grid): extracted aspect ratio calculation into seperate function

photodow committed:
feat(grid): updated aspect ratio grid to allow overflow

emyarod reviewed (commented):
it looks like a `prettier` error is causing CI to fail

photodow committed:
chore(grid): formatting sass

jeanservaas commented:
@joshblack 

@mjabbink  @photodow  and I hashed this out and have a better understanding of this I think. 

To clarify, the aspect ratio should only ever apply to tiles with limited type and images... there's no other reason people should be using the aspect ratio grid. However, it does seem like there's a clear need for a graceful overflow especially in edge cases involving accessibility. 

We were initially worried about situations like this causing vertical rag

![image](https://user-images.githubusercontent.com/43144260/73787009-ff5d6f00-475f-11ea-9ceb-fb0fc599027e.png)

But James preview shows that he has accounted for this and that at least all of the cards in the row can expand uniformly to prevent that.

![image](https://user-images.githubusercontent.com/43144260/73787086-2c118680-4760-11ea-8cb3-6c5aa8ee7b02.png)

We're also OK with people using the reverse of any of the aspect ratios that we have provided, i.e. 16 x 9, 9 x 16 etc.

jeanservaas reviewed (approved):
See my comments below.

joshblack commented:
Thanks for the update @jeanservaas! 

@photodow super curious, where does the `::before`, `::after` trick come from? Does it have a specific name?? Hadn't seen it before and wanted to learn more 👀

photodow commented:
@joshblack I found it via a typical trip down the rabbit hole.

Cool article on some of the future work on aspect ratios by Rachel Andrew
https://www.smashingmagazine.com/2019/03/aspect-ratio-unit-css/

She referred me to one of Chris Coyier's articles on aspect ratios and like all the different ways it can be done.
https://css-tricks.com/aspect-ratio-boxes/#article-header-id-6

I ultimately found his write up useful enough, but he also links out to some earlier articles on the topic from Keith Grant and even a 404 from Marc Hinse who apparently demoed it all the way back in 2013.
https://keithjgrant.com/posts/2017/03/aspect-ratios/

joshblack commented:
@photodow awesome, thanks so much for sharing! 🙏 Really appreciate it. Would you mind adding a comment to the code that uses this technique with a link to the article? Will help out a ton if we ever need to revisit this area of the codebase 👀

photodow committed:
chore(grid): adding css-tricks article

joshblack commented:
@photodow could you share more about removing `bx--aspect-ratio` and `bx--aspect-ratio--object`? We might need to keep these during the interim if we're moving towards this approach, just for breaking changes, but separately wanted to understand why we might not need them anymore. 

Would it make sense if this line: https://github.com/carbon-design-system/carbon/blob/1bf24e2bbf7d8b9cf38d85bef109bbcc5de8a50f/packages/grid/scss/_mixins.scss#L282 was applied to the `bx--aspect-ratio` or `bx--aspect-ratio--object` class?

Just for context, a lot of this code was borrowed from tachyons: https://github.com/tachyons-css/tachyons/blob/master/src/_aspect-ratios.css lol

photodow committed:
chore(grid): updated snapshots and apparently some other things

photodow commented:
> could you share more about removing bx--aspect-ratio and bx--aspect-ratio--object?

@joshblack I look at it as why have more HTML/CSS than necessary? `bx--aspect-ratio` could definitely still be used if we want to, but it felt excessive having `bx--aspect-ratio bx--aspect-ratio--6-9`, but I can see an argument made if you'd like to add that one back based on the practice of BEM possibly. The one that I'm struggling to see any value in is `bx--aspect-ratio--object`. It's a whole extra div that just doesn't provide any value (from what I can see) within this new technique.  If we're concerned about legacy, or breaking changes we could add the class in and add like a `height: 100%;` for people who might or might not be using it, but I don't see any reason to keep it for the aspect ratio. 🤷‍♂️That's how I was looking at it at least.

joshblack commented:
@photodow yeah, only meant it in terms of backwards-compat more than anything else!

For context for `bx--aspect-ratio--object`, it's meant for the immediate child versus having to be a whole extra node. In our examples I think we do encode it as an extra div, though 👍

joshblack commented:
Was trying to give this a thorough review today and I think the only breaking change I noticed was around the aspect ratio classes going from `bx--aspect-ratio--4x3` to `bx--aspect-ratio--4-3`

One QOL change that would be great is to update the attribute selector to `bx--aspect-ratio`. Would match our BEM conventions in this case and also reduces the footprint a little bit of the CSS bundle (albeit probably not by much lol).

In terms of `bx--aspect-ratio--object`, is it safe to drop it without a breaking change? Also noticed in the demo environment that the aspect ratios are on `bx--col`, instead would we want them as children of the column? When trying to do this locally I ran into the following issue and wasn't sure what to do:

![image](https://user-images.githubusercontent.com/3901764/74694080-50378380-51b4-11ea-9311-81f0f6238191.png)

Let me know and I can push up what I have or we can pair sometime tomorrow if you're free!

photodow committed:
chore(aspect-ratio): changed #-# back to #x#

photodow committed:
chore(aspect-ratio): seperated class names per BEM

photodow committed:
chore(aspect-ratio): updating example sass for height 100%

photodow committed:
chore(aspect-ratio): preventing breaking change for --object

photodow commented:
Any status on this? Anything I can do to help out?

joshblack commented on packages/grid/scss/_mixins.scss:
```diff
@@ -209,35 +209,95 @@
+/// @group @carbon/grid
+@mixin carbon--make-aspect-ratios($aspect-ratios: $carbon--aspect-ratios) {
+  .#{$prefix}--aspect-ratio {
+    @include carbon--make-aspect-ratio-ready;
```
Might make sense to inline this mixin for the selector. For the nested bit for the object, could unnest and write out the selector (follows our styleguide since we don't nest in Sass in this way)

joshblack commented on packages/grid/scss/_mixins.scss:
```diff
@@ -209,35 +209,95 @@
+      $width: nth($aspect-ratio, 1);
+      $height: nth($aspect-ratio, 2);
+
+      &--#{$width}x#{$height} {
```
Would be great to re-write this to avoid nesting

joshblack commented on packages/grid/scss/_mixins.scss:
```diff
@@ -209,35 +209,95 @@
+      $height: nth($aspect-ratio, 2);
+
+      &--#{$width}x#{$height} {
+        @include carbon--make-aspect-ratio($width, $height);
```
Should we just inline the call to the percentage since it's a one-liner?

joshblack reviewed (changes requested):
Thanks for making the updates!

photodow committed:
chore(aspect-ratio): removing nesting per style guidelines

photodow commented:
@joshblack I think I addressed all your feedback. Let me know if I misunderstood anything.

The one concern I have with these changes is it creates a dependency on these class names, and renders all possible ratios within carbon. If we only have one ratio that's needed to be used (e.g. 2x1) we'll still need to import all of them (e.g. 16x9, 9x16, 2x1, 1x2, 4x3, 3x4, 1x1).

joshblack commented:
@photodow could you call `carbon--make-aspect-ratios` like this to get only the ratios you want?

```scss
@include carbon--make-aspect-ratios((2, 1));
```

photodow commented:
> @photodow could you call carbon--make-aspect-ratios like this to get only the ratios you want?
> ```scss
> @include carbon--make-aspect-ratios((2, 1));
> ```

@joshblack I had a feeling you'd suggest that. :wink:

I thought about that too, but then you have the problem of potentially becoming out of sync with the Carbon core codebase. Might be able to leverage `map()`. Plus it makes responsive ratios difficult too.

We can consider tackling it in a separate issue though since this is about the graceful overflow specifically. Along with `@carbon/layout`.

joshblack committed:
Update _mixins.scss

joshblack commented:
cc @abbeyhrt let me know if you want to go over this for the review!

photodow committed:
Merge branch 'master' into aspect-ratio-overflow

photodow committed:
Merge branch 'master' into aspect-ratio-overflow

photodow commented:
@joshblack hey, any status or updates on this?

joshblack commented:
@photodow should be all good! I think just waiting on one more dev review, let me bug the team again 👀

abbeyhrt reviewed (approved):
Looks good to me! Thanks @photodow! 🎉

joshblack committed:
Merge branch 'master' into aspect-ratio-overflow

joshblack committed:
Merge branch 'master' into aspect-ratio-overflow

joshblack committed:
Merge branch 'master' into aspect-ratio-overflow

joshblack committed:
Merge branch 'master' into aspect-ratio-overflow

joshblack committed:
Merge branch 'master' into aspect-ratio-overflow
